src/VisualStudio/Core/Def/Implementation/FindUsages/RichNavFilterScopes.cs

```diff
@@ -0,0 +1,166 @@
+using Microsoft.Internal.VisualStudio.Shell.ErrorList;
+using Microsoft.VisualStudio.Shell.TableControl;
+using Microsoft.VisualStudio.Shell.TableManager;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindUsages
+{
+    internal static class RichNavOptions
+    {
+        internal const string RichNavAvailableOptionName = "RichNavAvailable";
+    }
+
+    [Export(typeof(IScopeFilterFactory))]
+    [TableManagerIdentifier("FindAllReferences*")]
+    [TableManagerIdentifier("FindResults*")]
+    [Replaces(PredefinedScopeFilterNames.EntireSolutionScopeFilter)]
+    [Replaces(PredefinedScopeFilterNames.AllItemsScopeFilter)]
+    [DeferCreation(OptionName = RichNavOptions.RichNavAvailableOptionName)] // This factory will not be loaded unless this option is set to Boolean true
+    [DefaultScope]
+    [Name(PredefinedScopeFilterNames.LoadedSolutionScopeFilter)]
+    [Order(Before = PredefinedScopeFilterNames.AllItemsScopeFilter)]
+    internal class LoadedSolutionScopeFilterFactory : IReplacingScopeFilterFactory
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public LoadedSolutionScopeFilterFactory()
+        {
+        }
+
+        public IErrorListFilterHandler CreateFilter(IWpfTableControl tableControl)
+        {
+            // We're only replacing an existing filter, and not creating a new one.
+            return null;
+        }
+
+        public IErrorListFilterHandler ReplaceFilter(IWpfTableControl tableControl, string filterIdentifier)
+        {
+            if (filterIdentifier == PredefinedScopeFilterNames.AllItemsScopeFilter)
+            {
+                return new LoadedSolutionFilterHandler(ServicesVSResources.Loaded_items);
+            }
+            else if (filterIdentifier == PredefinedScopeFilterNames.EntireSolutionScopeFilter)
+            {
+                return new LoadedSolutionFilterHandler(ServicesVSResources.Loaded_solution);
+            }
+
+            return null; // Don't replace
+        }
+    }
+
+    internal abstract class RoslynFilterHandler : FilterHandlerBase
+    {
+        private readonly ItemOrigin _origin;
+
+        protected RoslynFilterHandler(int id, string displayName, ItemOrigin origin)
+        {
+            FilterId = id;
+            FilterDisplayName = displayName;
+            _origin = origin;
+        }
+
+        public sealed override int FilterId { get; }
+
+        public sealed override string FilterDisplayName { get; }
+
+        public sealed override IEntryFilter GetFilter(out string displayText)
+        {
+            displayText = FilterDisplayName;
+            return new ItemOriginFilter(_origin);
+        }
+    }
+
+    internal class LoadedSolutionFilterHandler : RoslynFilterHandler
+    {
+        private const int LoadedSolutionFilterHandlerFilterId = 20;
```

Comment:
might be good to add a comment explaining why 20 is correct (same below).

Reply:
will do.  tehy were magic constants given to me is the answer :D

Reply:
this number must be different than ID of other filter handlers. There are existing handlers in the 1-19 range. This ID is mandated by a very old selection API.

Reply:
there is no list at the moment but I will add it to shell